vncviewer/fltk/Fl_Suggestion_Input.cxx

```diff
@@ -0,0 +1,711 @@
+        std::string before = full_label.substr(start, m.pos - start);
+        std::string match = full_label.substr(m.pos, m.len);
+
+        fl_font(orig_font, orig_size);
+        fl_draw(before.c_str(), x + offset, y);
+        offset += fl_width(before.c_str());
+
+        fl_font(FL_BOLD, orig_size);
+        fl_draw(match.c_str(), x + offset, y);
+        offset += fl_width(match.c_str());
+
+        start = m.pos + m.len;
+    }
+
+    std::string after = full_label.substr(start);
+
+    fl_font(orig_font, orig_size);
+    fl_draw(after.c_str(), x + offset, y);
+    offset += fl_width(after.c_str());
+}
+
+void Fl_Suggestion_Item::emphasize_match(const std::string match)
+{
+    matches.clear();
+    if (match == "")
+        return;
+
+    std::string _match = on_normalize_cb(match);
+    std::string _full_label = on_normalize_cb(full_label);
+
+    std::list<std::string> words = split_on_space(_match);
+    if (words.empty())
+        words.push_back(_match);
+
+    for (std::string w : words) {
+        std::string::size_type pos = _full_label.find(w);
+        if (w != "" && pos != std::string::npos) {
+            Match m;
+            m.pos = pos;
+            m.len = w.length();
+            matches.push_back(m);
+        }
+    }
+    redraw();
+}
+
+void Fl_Suggestion_Item::select_cb(Fl_Widget *sb, void *w)
+{
+    Fl_Button * select_button = (Fl_Button*) sb;
+    Fl_Suggestion_Item * item = (Fl_Suggestion_Item*) select_button->parent();
+    Fl_Suggestion_Input * widget = (Fl_Suggestion_Input*) w;
+    widget->select_suggestion(item->label());
+    widget->hide_suggestions();
+}
+void Fl_Suggestion_Item::remove_cb(Fl_Widget *rb, void *w)
+{
+    Fl_Button * remove_button = (Fl_Button*) rb;
+    Fl_Suggestion_Item * item = (Fl_Suggestion_Item*) remove_button->parent();
+    Fl_Suggestion_Input * widget = (Fl_Suggestion_Input*) w;
+    widget->remove_suggestion(item->label());
+}
+
+/*
+ * ============== Fl_Suggestion_Input ==============
+ */
+
+Fl_Suggestion_Input::Fl_Suggestion_Input(
+    int x, int y, int w, int h, const char *label,
+    const std::list<std::string> &_suggestions
+)
+: Fl_Group(x, y, w, h, label), suggestions(_suggestions),
+  searching(false)
+{
+    Fl_Group::color(FL_BACKGROUND2_COLOR);
+    Fl_Group::box(FL_DOWN_BOX);
+    align(FL_ALIGN_LEFT); // Place label outside, on the left side
+
+    input = new Fl_Input(x + INPUT_OFFSET, y + INPUT_OFFSET,
+                         w - INPUT_OFFSET - 5,
+                         h - (INPUT_OFFSET * 2));
+    input->box(FL_FLAT_BOX);
+    input->color(FL_BACKGROUND2_COLOR);
+    input->callback(text_changed_cb, this);
+    input->when(FL_WHEN_CHANGED);
+
+    max_width = w - 2 * SUGGESTION_OFFSET;
+    y += h;
+
+    end(); // Close Fl_Group
+
+    // The dropdown window is not a child of Fl_Suggestion_Input, but will
+    // instead be a sibling. Before being shown it is added to the parent.
+    dropdown_window = new Fl_Window(x, y, w, 0);
```

Comment:
This does not seem to work correctly. I'm not getting a new window, but rather just a widget container inside the existing window.